File: matrix_mult/matrix_mult.c
#include <time.h>
#include <err.h>
#include <linux/perf_event.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/syscall.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdint.h>
#include <inttypes.h>

#define TOTAL_EVENTS 5

static long perf_event_open(struct perf_event_attr *hw_event, pid_t pid,
                            int cpu, int group_fd, unsigned long flags)
{
    return syscall(SYS_perf_event_open, hw_event, pid, cpu, group_fd, flags);
}

void configure_event(struct perf_event_attr *pe, uint32_t type, uint64_t config)
{
    memset(pe, 0, sizeof(struct perf_event_attr));
    pe->type = type;
    pe->size = sizeof(struct perf_event_attr);
    pe->config = config;
    pe->read_format = PERF_FORMAT_GROUP | PERF_FORMAT_ID;
    pe->disabled = 1;
    pe->exclude_kernel = 1;
    pe->exclude_hv = 1;
}

struct read_format
{
    uint64_t nr;
    struct
    {
        uint64_t value;
        uint64_t id;
    } values[TOTAL_EVENTS];
};

double diff_timespec(struct timespec *time1, struct timespec *time0)
{
    return (time1->tv_sec - time0->tv_sec) + (time1->tv_nsec - time0->tv_nsec) / 1000000000.0;
}

int **allocate_matrix(int size)
{
    int **matrix = (int **)malloc(size * sizeof(int *));
    if (!matrix)
    {
        printf("Error: Memory allocation failed.\n");
        exit(1);
    }
    for (int i = 0; i < size; i++)
    {
        matrix[i] = (int *)malloc(size * sizeof(int));
        if (!matrix[i])
        {
            printf("Error: Memory allocation failed.\n");
            exit(1);
        }
    }
    return matrix;
}

void free_matrix(int **matrix, int size)
{
    for (int i = 0; i < size; i++)
    {
        free(matrix[i]);
    }
    free(matrix);
}

void fill_matrix(int **matrix, int size)
{
    for (int i = 0; i < size; i++)
    {
        for (int j = 0; j < size; j++)
        {
            matrix[i][j] = rand();
        }
    }
}

void multiply_matrices(int **A, int **B, int **C, int size)
{
    for (int i = 0; i < size; i++)
    {
        for (int j = 0; j < size; j++)
        {
            C[i][j] = 0;
            for (int k = 0; k < size; k++)
            {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
}

void print_matrix(int **matrix, int size, const char *name)
{
    printf("Matrix %s:\n", name);
    for (int i = 0; i < size; i++)
    {
        for (int j = 0; j < size; j++)
        {
            printf("%5d ", matrix[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}

int main(int argc, char *argv[])
{
    if (argc != 2)
    {
        printf("Usage: %s <matrix_size>\n", argv[0]);
        return 1;
    }

    int size = atoi(argv[1]);
    if (size <= 0)
    {
        printf("Error: matrix size must be a positive integer.\n");
        return 1;
    }

    struct timespec start_time, end_time;
    srand(10000);

    // Allocate matrices
    int **A = allocate_matrix(size);
    int **B = allocate_matrix(size);
    int **C = allocate_matrix(size);

    fill_matrix(A, size);
    fill_matrix(B, size);

    // print_matrix(A, size, "A");
    int fd[TOTAL_EVENTS];
    int id[TOTAL_EVENTS];
    uint64_t pe_val[TOTAL_EVENTS];
    struct perf_event_attr pe[TOTAL_EVENTS];
    struct read_format counter_results;

    // Configure desired events
    configure_event(&pe[0], PERF_TYPE_HARDWARE, PERF_COUNT_HW_CPU_CYCLES);
    configure_event(&pe[1], PERF_TYPE_HARDWARE, PERF_COUNT_HW_INSTRUCTIONS);
    configure_event(&pe[2], PERF_TYPE_HW_CACHE,
                    (PERF_COUNT_HW_CACHE_L1D | (PERF_COUNT_HW_CACHE_OP_READ << 8) |
                     (PERF_COUNT_HW_CACHE_RESULT_MISS << 16)));
    configure_event(&pe[3], PERF_TYPE_HW_CACHE,
                    (PERF_COUNT_HW_CACHE_LL | (PERF_COUNT_HW_CACHE_OP_READ << 8) |
                     (PERF_COUNT_HW_CACHE_RESULT_MISS << 16)));
    configure_event(&pe[4], PERF_TYPE_HARDWARE, PERF_COUNT_HW_BRANCH_MISSES);

    // Open group leader
    fd[0] = perf_event_open(&pe[0], 0, -1, -1, 0);
    ioctl(fd[0], PERF_EVENT_IOC_ID, &id[0]);

    // Open other events as part of the group
    for (int i = 1; i < TOTAL_EVENTS; i++)
    {
        fd[i] = perf_event_open(&pe[i], 0, -1, fd[0], 0);
        ioctl(fd[i], PERF_EVENT_IOC_ID, &id[i]);
    }

    ioctl(fd[0], PERF_EVENT_IOC_RESET, PERF_IOC_FLAG_GROUP);
    ioctl(fd[0], PERF_EVENT_IOC_ENABLE, PERF_IOC_FLAG_GROUP);

    clock_gettime(CLOCK_MONOTONIC, &start_time);
    multiply_matrices(A, B, C, size);
    clock_gettime(CLOCK_MONOTONIC, &end_time);

    ioctl(fd[0], PERF_EVENT_IOC_DISABLE, PERF_IOC_FLAG_GROUP);
    read(fd[0], &counter_results, sizeof(struct read_format));

    printf("Num events captured: %" PRIu64 "\n", counter_results.nr);
    for (int i = 0; i < counter_results.nr; i++)
    {
        for (int j = 0; j < TOTAL_EVENTS; j++)
        {
            if (counter_results.values[i].id == id[j])
            {
                pe_val[j] = counter_results.values[i].value;
            }
        }
    }

    printf("CPU cycles:             %" PRIu64 "\n", pe_val[0]);
    printf("Instructions retired:   %" PRIu64 "\n", pe_val[1]);
    printf("L1 DCache read misses:   %" PRIu64 "\n", pe_val[2]);
    printf("Last level DCache read misses:   %" PRIu64 "\n", pe_val[3]);
    printf("Branch misses:           %" PRIu64 "\n", pe_val[4]);
    printf("IPC (instructions/cycle): %.2f\n", (double)pe_val[1] / pe_val[0]);

    for (int i = 0; i < TOTAL_EVENTS; i++)
    {
        close(fd[i]);
    }

    double time_elapsed = diff_timespec(&end_time, &start_time);
    printf("Time taken to multiply %dx%d matrices: %f seconds\n", size, size, time_elapsed);

    free_matrix(A, size);
    free_matrix(B, size);
    free_matrix(C, size);

    return 0;
}

File: matrix_mult/src/matrix_mult.rs
use libc::c_int;
use libc::c_long;
use libc::c_ulong;
use libc::pid_t;
use libc::SYS_perf_event_open;
use libc::{clock_gettime, rand, srand, timespec, CLOCK_MONOTONIC};
use perf_event_open_sys::bindings::perf_event_attr;
use perf_event_open_sys::{bindings, ioctls};
use std::env;
use std::io::Read;
use std::mem;
use std::process::exit;
use std::{fs::File, os::unix::io::FromRawFd};

const TOTAL_EVENTS: usize = 5;

unsafe fn perf_event_open(
    hw_event: *mut perf_event_attr,
    pid: pid_t,
    cpu: c_int,
    group_fd: c_int,
    flags: c_ulong,
) -> c_long {
    libc::syscall(SYS_perf_event_open, hw_event, pid, cpu, group_fd, flags)
}
fn configure_event(attr: &mut bindings::perf_event_attr, type_: u32, config: u64) {
    *attr = bindings::perf_event_attr::default();
    attr.type_ = type_;
    attr.size = mem::size_of::<bindings::perf_event_attr>() as u32;
    attr.config = config;
    attr.read_format = (bindings::PERF_FORMAT_GROUP | bindings::PERF_FORMAT_ID) as u64;
    attr.set_disabled(1);
    attr.set_exclude_kernel(1);
    attr.set_exclude_hv(1);
}

#[derive(Default, Copy, Clone)]
struct ReadValue {
    value: u64,
    id: u64,
}

#[derive(Default)]
struct ReadFormat {
    nr: u64,
    values: [ReadValue; TOTAL_EVENTS],
}

fn allocate_matrix(size: usize) -> Vec<Vec<i32>> {
    vec![vec![0; size]; size]
}

fn fill_matrix(matrix: &mut Vec<Vec<i32>>, size: usize) {
    for i in 0..size {
        for j in 0..size {
            matrix[i][j] = unsafe { rand() }
        }
    }
}

fn multiply_matrices(
    matrix_a: &Vec<Vec<i32>>,
    matrix_b: &Vec<Vec<i32>>,
    matrix_result: &mut Vec<Vec<i32>>,
    size: usize,
) {
    for i in 0..size {
        for j in 0..size {
            matrix_result[i][j] = 0;
            for k in 0..size {
                matrix_result[i][j] += matrix_a[i][k] * matrix_b[k][j];
            }
        }
    }
}

fn diff_timespec(time1: &timespec, time0: &timespec) -> f64 {
    (time1.tv_sec - time0.tv_sec) as f64 + (time1.tv_nsec - time0.tv_nsec) as f64 / 1_000_000_000.0
}

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() != 2 {
        eprintln!("Usage: {} <matrix_size>", args[0]);
        exit(1);
    }

    let size: usize = match args[1].parse() {
        Ok(n) if n > 0 => n,
        _ => {
            eprintln!("Error: matrix size must be a positive integer.");
            exit(1);
        }
    };

    unsafe {
        srand(10000);
    }

    let mut matrix_a = allocate_matrix(size);
    let mut matrix_b = allocate_matrix(size);
    let mut matrix_result = allocate_matrix(size);

    fill_matrix(&mut matrix_a, size);
    fill_matrix(&mut matrix_b, size);
    let mut attrs = vec![perf_event_attr::default(); TOTAL_EVENTS];
    let mut fds: [i32; TOTAL_EVENTS] = [-1; TOTAL_EVENTS];
    let mut ids = [0u64; TOTAL_EVENTS];
    let mut pe_val = [0u64; TOTAL_EVENTS];

    configure_event(
        &mut attrs[0],
        bindings::PERF_TYPE_HARDWARE,
        bindings::PERF_COUNT_HW_CPU_CYCLES as u64,
    );
    configure_event(
        &mut attrs[1],
        bindings::PERF_TYPE_HARDWARE,
        bindings::PERF_COUNT_HW_INSTRUCTIONS as u64,
    );
    configure_event(
        &mut attrs[2],
        bindings::PERF_TYPE_HW_CACHE,
        (bindings::PERF_COUNT_HW_CACHE_L1D as u64)
            | ((bindings::PERF_COUNT_HW_CACHE_OP_READ as u64) << 8)
            | ((bindings::PERF_COUNT_HW_CACHE_RESULT_MISS as u64) << 16),
    );
    configure_event(
        &mut attrs[3],
        bindings::PERF_TYPE_HW_CACHE,
        (bindings::PERF_COUNT_HW_CACHE_LL as u64)
            | ((bindings::PERF_COUNT_HW_CACHE_OP_READ as u64) << 8)
            | ((bindings::PERF_COUNT_HW_CACHE_RESULT_MISS as u64) << 16),
    );
    configure_event(
        &mut attrs[4],
        bindings::PERF_TYPE_HARDWARE,
        bindings::PERF_COUNT_HW_BRANCH_MISSES as u64,
    );

    unsafe {
        fds[0] = perf_event_open(&mut attrs[0], 0, -1, -1, 0) as i32;
        ioctls::ID(fds[0], &mut ids[0]);

        for i in 1..TOTAL_EVENTS {
            fds[i] = perf_event_open(&mut attrs[i], 0, -1, fds[0], 0) as i32;
            ioctls::ID(fds[i], &mut ids[i]);
        }

        ioctls::RESET(fds[0], bindings::PERF_IOC_FLAG_GROUP);
        ioctls::ENABLE(fds[0], bindings::PERF_IOC_FLAG_GROUP);
    }

    let mut start_time: timespec = timespec {
        tv_sec: 0,
        tv_nsec: 0,
    };
    unsafe {
        clock_gettime(CLOCK_MONOTONIC, &mut start_time);
    }

    multiply_matrices(&matrix_a, &matrix_b, &mut matrix_result, size);

    let mut end_time: timespec = timespec {
        tv_sec: 0,
        tv_nsec: 0,
    };
    unsafe {
        clock_gettime(CLOCK_MONOTONIC, &mut end_time);
    }
    unsafe {
        ioctls::DISABLE(fds[0], bindings::PERF_IOC_FLAG_GROUP);
    }
    let mut buf = ReadFormat::default();
    let mut file = unsafe { File::from_raw_fd(fds[0]) };
    unsafe {
        let raw_buf = std::slice::from_raw_parts_mut(
            &mut buf as *mut _ as *mut u8,
            mem::size_of::<ReadFormat>(),
        );
        file.read_exact(raw_buf).unwrap();
    }

    for i in 0..buf.nr as usize {
        for j in 0..TOTAL_EVENTS {
            if buf.values[i].id == ids[j] {
                pe_val[j] = buf.values[i].value;
            }
        }
    }

    println!("CPU cycles:             {}", pe_val[0]);
    println!("Instructions retired:   {}", pe_val[1]);
    println!("L1 DCache read misses:   {}", pe_val[2]);
    println!("Last level DCache read misses:   {}", pe_val[3]);
    println!("Branch misses:           {}", pe_val[4]);
    println!(
        "IPC (instructions/cycle): {:.2}",
        pe_val[1] as f64 / pe_val[0] as f64
    );
    for fd in fds {
        unsafe {
            libc::close(fd.try_into().unwrap());
        }
    }

    let time_elapsed = diff_timespec(&end_time, &start_time);

    println!(
        "Time taken to multiply {}x{} matrices: {:.6} seconds",
        size, size, time_elapsed
    );
}
